### ISSUE TITLE
Understand Jupyterhub deployment for causal analysis

### ISSUE THREAD

**Issue by Shreyanand** [aicoe-aiops/operate-first-openshift-data#18]
As a first step of the new project with IBM, we'd like to understand the structure of pods and their relationships in a deployment. To start with a simple example, we want to look at all the pods in the Jupyterhub deployment. 
Currently, in the Openshift [console](https://console-openshift-console.apps.zero.massopen.cloud/k8s/ns/opf-jupyterhub/pods), we see two pods that seem to be "parent" pods (jupyterhub-123, jupyter-db-2) and then 22 other pods, each for one user of the JH instance. 
There are a few questions about this that we would like to answer:

1. Is that all that there is to this deployment? If yes, how does the communication work here? How are the 22 pods dependent on the 2 pods? What else do the 2 pods do? Are there more pods that these pods communicate with? 
2. The [jupyterhub metrics](http://thanos-query-frontend-opf-observatorium.apps.zero.massopen.cloud/graph?g0.range_input=1h&g0.max_source_resolution=0s&g0.expr=jupyterhub_proxy_add_duration_seconds_sum&g0.tab=1)  seem to be all from the jupyterhub-123 pod. How are these [metrics](https://grafana-route-opf-monitoring.apps.zero.massopen.cloud/d/YuYfkHYMk/jupyterhub-usage?orgId=1&var-user_id=shanand-40redhat-2ecom) computed?
3. We don't seem to have access to everything in the openshift console. Can we get it to know more about the PVCs, routes, services, etc. in order to get a more complete picture of how everything happens? Or we could set up some time with someone who does have access.

The aim of this exercise would be to identify all the stakeholder elements in the example so that we can then extract logs, metrics, events for them. This could potentially serve as ground truth for validating a learned structure.

**Comment by Shreyanand** [comment]
cc @tumido @4n4nd Any pointers for this?

**Comment by MichaelClifford** [comment]
> 2) How are these metrics computed?

You can use the "explore" function on these graphs and see what Prometheus metrics are being used to compute them. 

![image](https://user-images.githubusercontent.com/4494906/114754604-c1c28180-9d26-11eb-982a-6a9da0881e32.png)

The "explore" dashboard provides the promQL that defines what displayed on the graph. 

![image](https://user-images.githubusercontent.com/4494906/114754732-e3236d80-9d26-11eb-960c-673a6d871cbe.png)


All of these  metrics are defined by the following two labels `{namespace="opf-jupyterhub", pod="prometheus-k8s-0"}`

**Comment by Shreyanand** [comment]
> > 1. How are these metrics computed?
> 
> You can use the "explore" function on these graphs and see what Prometheus metrics are being used to compute them.
> 
> ![image](https://user-images.githubusercontent.com/4494906/114754604-c1c28180-9d26-11eb-982a-6a9da0881e32.png)
> 
> The "explore" dashboard provides the promQL that defines what displayed on the graph.
> 
> ![image](https://user-images.githubusercontent.com/4494906/114754732-e3236d80-9d26-11eb-960c-673a6d871cbe.png)
> 
> All of these metrics are defined by the following two labels `{namespace="opf-jupyterhub", pod="prometheus-k8s-0"}`

Cool! so for CPU usage it's container_cpu_usage_seconds_total, for memory usage it's container_memory_working_set_bytes, and for PVC usage it's kubelet_volume_stats_capacity_bytes. Grouped by the pod names ofcourse.

**Comment by tumido** [comment]
I think this image describes the architecture the best. In our OpenShift/ODH-powered case the pods are named a bit differently but in general the mapping is still valid:

![](https://zero-to-jupyterhub.readthedocs.io/en/stable/_images/architecture.png)
https://zero-to-jupyterhub.readthedocs.io/en/stable/administrator/architecture.html

The `proxy-*` and `hub-*` boxes are running within the `jupyterhub-*` pod. This pod is running image build from [this repo](https://github.com/opendatahub-io/jupyterhub-odh) which is based on a builder image from [this repo](https://github.com/opendatahub-io/jupyterhub-quickstart).

It communicates with its db pod, which is just a generic postgresql for user-related data as described [here](https://jupyterhub.readthedocs.io/en/stable/reference/database.html)
 
Then there's the user pods which are accessed through the proxy running within the `jupyterhub-*` pod. And that's basically it.

The proxy is creating most of the logs in the `jupyterhub-*` pod. The pods uses a service account to query for available imagestreams and single user profiles and it is also used to schedule and stop the user pods.

**Comment by Shreyanand** [comment]
Thanks @tumido! Based on your answer, a discussion with @4n4nd, and my understanding of it all, I made this rough draft of the pod structure:
![jupyterhub_structure (1)](https://user-images.githubusercontent.com/8916126/115425211-0bc2d000-a21d-11eb-85a6-f989b3478918.png)


- The ovals represent a pod
- Green color means that we can observe logs, metrics, and events data for these pods, no color indicates it may be difficult to obtain those. 
- The arrows represent the causal flow (It may need a closer look)

Explanation of what each pod is doing:

1. The most important one seems to be the jupyterhub-* pod. The suffix here is usually the deployment number (so in the current case it's 123) followed by some identifier. This pod has a service account and is responsible for authentication and spawning user pods. When a user tries to start a server, first this pod communicates with the Openshift Oauth pod(s) to validate if the user has right credentials. If that is the case, it then communicates with Openshift API to allocate the desired resources (RAM, CPU, PVC, GPU) to the user and create a pod for them. The arrows go from Openshift API to jupyterhub-* since a problem with the former can give errors to the latter and not the other way round. 

2. jupyterhub-db pod is a database pod that stores user information like user details, session details etc. 

3. There is a third pod called "spark-operator" that spawns master and worker pods for a user if they specify that they need a spark cluster for their application.  

Questions: 

- Is there a pod for Openshift API? (that lets Jpuyterhub to create user pods)

- Who tells spark operator pod  when to create master and worker pods? The jupyterhub-* pod or a user pod? Does it also have a service account or does it create new nodes through the jupyterhub-* pod? This shall help us connect the spark pods to the rest of the pods.

**Comment by MichaelClifford** [comment]
@Shreyanand @4n4nd  for the spark side diagram here, Shouldn't there be a line from the green oval for user 1 to the master spark pod for user 1?  As instructions (possibly ones that can crash pods) are sent in that direction?

Also, I wonder if including Spark in the equation makes sense just yet as it is an optional and only sometimes used feature that complicates the problem space a bit. I would recommend getting pure jupyterhub nailed down, then extending to the spark cases as step 2. WDYT?

**Comment by MichaelClifford** [comment]
@Shreyanand there is no entry in "Explanation of what each pod is doing:" section for the user pods. Should that be added as well?

**Comment by Shreyanand** [comment]
> @Shreyanand @4n4nd for the spark side diagram here, Shouldn't there be a line from the green oval for user 1 to the master spark pod for user 1? As instructions (possibly ones that can crash pods) are sent in that direction?

It could be that. Although I'm not sure. 

> Also, I wonder if including Spark in the equation makes sense just yet as it is an optional and only sometimes used feature that complicates the problem space a bit. I would recommend getting pure jupyterhub nailed down, then extending to the spark cases as step 2. WDYT?

Right, we could totally ignore that bit for the initial iteration.

**Comment by tumido** [comment]
> Is there a pod for Openshift API? (that lets Jpuyterhub to create user pods)

Yes, OpenShift API lives in `openshift-apiserver` namespace. There are 3 replicas (pods) in the deployment (1 pod per master node). The API server is managed by the cluster operator for API which lives in `openshift-apiserver-operator`.

Similarly the Auth lives in `openshift-authentication` in 2 replicas. Again, this is managed via a cluster operator living in `openshift-authentication-operator`.

> Who tells spark operator pod when to create master and worker pods?

This is specified in the profile [here](https://github.com/opendatahub-io/odh-manifests/blob/171b3d5399e25b40ca675bb48eecbcf926aafb7b/jupyterhub/jupyterhub/base/jupyterhub-singleuser-profiles-configmap.yaml#L20-L56). This instructs the `jupyterhub-*` controller pod to apply specific settings (env variables) when spawning the JH images listed there + request a spark cluster to be created. This Spark cluster is created by submitting [this manifest](https://github.com/opendatahub-io/odh-manifests/blob/171b3d5399e25b40ca675bb48eecbcf926aafb7b/jupyterhub/jupyterhub/base/jupyterhub-spark-operator-configmap.yaml#L20-L41). The `SparkCluster` resource is consumed by the Spark operator which then creates the cluster in requested configuration. 

Spark operator lives in the same namespace as JH. It's called `spark-operator` deployment (`spark-operator-*` pod)
